core/server/worker/src/main/java/alluxio/worker/grpc/BlockReadHandler.java

```diff
@@ -540,7 +541,16 @@ protected DataBuffer getDataBuffer(BlockReadRequestContext context, long offset,
           case FILE:
             //TODO(beinan): change the blockReader interface to accept pre-allocated byte buffer
             // or accept a supplier of the bytebuffer.
-            if (mIsReaderBufferPooled) {
+            if (mIsNioBufferPoolEnabled) {
+              buf = PooledDirectNioByteBuf.allocate(len);
+              try {
+                while (buf.writableBytes() > 0 && blockReader.transferTo(buf) != -1) {
+                }
+                return new NettyDataBuffer(buf.retain());
+              } finally {
+                buf.release();
+              }
+            } else if (mIsReaderBufferPooled) {
```

Comment:
judging from the name, shouldn't `mIsReaderBufferPooled` has priority over mIsNioBufferPoolEnabled? i.e. if `mIsReaderBufferPooled` is false, then neither the NIO buffer pool nor Netty pool should be used, so the buffer is not pooled at all.

Reply:
Interesting question, so we should judge the mIsReaderBufferPooled first, something like

```java

 if (mIsReaderBufferPooled) {
  buf = mIsNioBufferPoolEnabled ? PooledDirectNioByteBuf.allocate(len) : PooledByteBufAllocator.DEFAULT.buffer(len, len);
  try {
    while (buf.writableBytes() > 0 && blockReader.transferTo(buf) != -1) {
    }
    return new NettyDataBuffer(buf.retain());
  } finally {
    buf.release();
  }
} else {
  ByteBuffer buffer = blockReader.read(offset, len);
  return new NettyDataBuffer(Unpooled.wrappedBuffer(buffer));
}
```

Reply:
I might combine mIsReaderBufferPooled and mIsNioBufferPoolEnabled once we figure out a best practice of pooling buffer, as Bowen said, actually there are just two behavior, pooled or not pooled,

Reply:
Why do we need to have a `buf.retain()` and then execute `buf.release()` in the finally block? Is it to release immediately when an exception occurs before executing `buf.retain()`?

Reply:
> release immediately when an exception occurs before executing buf.retain()?

The `finally` block is there to ensure the buffer is correctly released if an exception occurs during `blockReader.transferTo`, so that it is not leaked (in the sense that if it's a pooled buffer, it's not returned to the pool). But if no exception occurs, we then hand over the ownership of the buffer to `NettyDataBuffer`, thus need to increment the reference count before we release it in the finally block.

I think it may be less confusing if it's structured this way:

```java
  try {
    while (buf.writableBytes() > 0 && blockReader.transferTo(buf) != -1) {
    }
  } catch (Throwable e) {
    buf.release();
    throw e;
  }
  return new NettyDataBuffer(buf);
```

Reply:
This is much clearer.